### PR TITLE
BTC Signing reorg

### PR DESCRIPTION
**Pull request by catenocrypt**
## Description

Refactoring of BTC Signing:
- Reduce usage of Proto:: classes internally, some classes had their corresponding non-proto class, but not all.
- Named classes/aliases for TransactionInputs, Outputs, UTXOs, instead of plan vector<>
- Better separation of tx building/planning/signing phases.
- Tests adaptions to not use Proto classes (a bit more readable).

## Testing instructions

Unit tests.

## Types of changes

* Non-breaking internal refactoring

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain
- [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).

### CODE REVIEW

**Comment by hewigovens** [comment]
@catenocrypt since we're changing bitcoin utxo classes, could we add virtual UTXO class and make `UnspentSelector` generic? basically is this issue: https://github.com/trustwallet/wallet-core/issues/1495, this will reduce future utxo chain integration effort

**Comment by catenocrypt** [comment]
> @catenocrypt since we're changing bitcoin utxo classes, could we add virtual UTXO class and make `UnspentSelector` generic? basically is this issue: #1495, this will reduce future utxo chain integration effort

I'll check what's the best way to generalize.  My reorg is mostly done, I'll see if it makes sense as part of this PR or a different one.

**Comment by catenocrypt** [comment]
> > @catenocrypt since we're changing bitcoin utxo classes, could we add virtual UTXO class and make `UnspentSelector` generic? basically is this issue: #1495, this will reduce future utxo chain integration effort
> 
> I'll check what's the best way to generalize. My reorg is mostly done, I'll see if it makes sense as part of this PR or a different one.

The UnspentSelector covers the currently existing UTXO chains.   If need be, it can be easily made more generic (e.g. template).  FeeSelector is also quite generic.  I don't add generic UTXO handling with this PR (e.g. on proto level).

**Comment by hewigovens** [comment]
Ok, let's do it in a separate pr

File: src/Bitcoin/TransactionSigner.h [hunk 1]
```diff
@@ -8,75 +8,69 @@
...
 
 namespace TW::Bitcoin {
 
-/// Helper class that performs Bitcoin transaction signing.
+/// Frontend class for transaction planning, building, and signing
 template <typename Transaction, typename TransactionBuilder>
 class TransactionSigner {
-  private:
+public:
+    // Create plan for a transaction
```
Comment (vikmeup) [inline src/Bitcoin/TransactionSigner.h:28]: is this TODO for this PR to tackle?

File: src/Bitcoin/TransactionSigner.h [hunk 2]
```diff
@@ -8,75 +8,69 @@
...
+    // TODO rename
+    static TransactionPlan plan(const SigningInput& input);
+
+    // Sign an unsigned transaction.  Plan it if needed beforehand.
+    static Result<Transaction, Common::Proto::SigningError> sign(const SigningInput& input, bool estimationMode = false);
+};
+
+/// Class that performs Bitcoin transaction signing.
```
Comment (vikmeup) [inline src/Bitcoin/TransactionSigner.h:36]: is this TODO for this PR to tackle?

**Review by vikmeup — APPROVED** [review]

File: src/Bitcoin/SigningInput.h [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+#include <string>
+#include <vector>
+#include <map>
+#include <optional>
+
+namespace TW::Bitcoin {
+
+/// Input for signing, info of an unsigned transaction
+class SigningInput {
```
Comment (hewigovens) [inline src/Bitcoin/SigningInput.h:23]: I guess we can break signing input into smaller classes? group by functionalities etc, no limit like the single proto model

File: src/Bitcoin/TransactionBuilder.cpp [hunk 4]
```diff
@@ -15,28 +16,36 @@
 
 namespace TW::Bitcoin {
 
+std::optional<TransactionOutput> TransactionBuilder::prepareOutput(std::string address, Amount amount, enum TWCoinType coin) {
```
Comment (hewigovens) [inline src/Bitcoin/TransactionBuilder.cpp:19]:
```suggestion
std::optional<TransactionOutput> TransactionBuilder::prepareOutputWithScript(std::string address, Amount amount, enum TWCoinType coin) {
```

**Review by hewigovens — APPROVED** [review]

File: src/Bitcoin/TransactionSigner.h (hunk 2 above)
Comment (catenocrypt) [reply src/Bitcoin/TransactionSigner.h]: separation into 2 source files done

**Review by vikmeup — APPROVED** [review]

**Pull request merged**